File: src/cs/Ssh/Services/ConnectionService.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Events;
using Microsoft.DevTunnels.Ssh.Messages;

namespace Microsoft.DevTunnels.Ssh.Services;

/// <summary>
/// Handles SSH protocol messages related to opening channels.
/// </summary>
[ServiceActivation(ServiceRequest = Name)]
#pragma warning disable CA1812 // Avoid Uninstantiated Internal Classes
internal class ConnectionService : SshService
#pragma warning restore CA1812 // Avoid Uninstantiated Internal Classes
{
	/// <summary>
	/// Tracks a channel open request that has been sent and not yet acknowledged.
	/// Used as a value in the 'pendingChannels' dictionary; the channel ID is the key.
	/// </summary>
	private class PendingChannel
	{
		public PendingChannel(
			 ChannelOpenMessage openMessage,
			 TaskCompletionSource<SshChannel> completionSource)
		{
			OpenMessage = openMessage;
			CompletionSource = completionSource;
		}

		public ChannelOpenMessage OpenMessage { get; }
		public TaskCompletionSource<SshChannel> CompletionSource { get; }
		public CancellationTokenRegistration CancellationRegistration { get; set; }
	}

	public const string Name = "ssh-connection";

	private readonly object lockObject = new object();
	private readonly IDictionary<uint, SshChannel> channels;
	private readonly IDictionary<uint, SshChannel> nonAcceptedChannels;
	private readonly IDictionary<uint, PendingChannel> pendingChannels;
	private readonly IDictionary<string, IList<TaskCompletionSource<SshChannel>>> pendingAcceptChannels;
	private bool disposed;
	private long channelCounter = -1;
	private Exception? closedException = null;

	public ConnectionService(SshSession session) : base(session)
	{
		this.channels = new Dictionary<uint, SshChannel>();
		this.nonAcceptedChannels = new Dictionary<uint, SshChannel>();
		this.pendingChannels = new Dictionary<uint, PendingChannel>();
		this.pendingAcceptChannels =
			new Dictionary<string, IList<TaskCompletionSource<SshChannel>>>();
	}

	public ICollection<SshChannel> Channels => this.channels.Values;

	public void Close(Exception ex)
	{
		this.closedException = ex;

		SshChannel[] channels;
		TaskCompletionSource<SshChannel>[] channelCompletions;
		lock (this.lockObject)
		{
			channels = Channels.ToArray();
			channelCompletions = this.pendingChannels.Select((c) => c.Value.CompletionSource)
				.Concat(this.pendingAcceptChannels.SelectMany(pac => pac.Value)).ToArray();
		}

		foreach (var channel in channels)
		{
			channel.Close(ex);
		}

		foreach (var channelCompletion in channelCompletions)
		{
			channelCompletion.TrySetException(ex);
		}
	}

	protected override void Dispose(bool disposing)
	{
		if (disposing)
		{
			SshChannel[] channels;
			TaskCompletionSource<SshChannel>[] channelCompletions;
			lock (this.lockObject)
			{
				channels = Channels.ToArray();
				channelCompletions = this.pendingChannels.Select((c) => c.Value.CompletionSource)
					.Concat(this.pendingAcceptChannels.SelectMany(pac => pac.Value)).ToArray();
				this.disposed = true;
			}

			foreach (var channel in channels)
			{
				channel.Dispose();
			}

			foreach (var channelCompletion in channelCompletions)
			{
				channelCompletion.TrySetException(
					new ObjectDisposedException(Session.ToString(), "Session closed."));
			}
		}

		base.Dispose(disposing);
	}

	internal async Task<uint> OpenChannelAsync(
		ChannelOpenMessage openMessage,
		TaskCompletionSource<SshChannel> completionSource,
		CancellationToken cancellation)
	{
		uint channelId = (uint)Interlocked.Increment(ref this.channelCounter);
		if (channelId > int.MaxValue)
		{
			this.channelCounter = int.MaxValue;
			throw new InvalidOperationException("Maximum number of channels reached.");
		}

		openMessage.SenderChannel = channelId;

		lock (this.lockObject)
		{
			if (this.disposed)
			{
				throw this.closedException ??
					new ObjectDisposedException(Session.ToString(), "Session closed.");
			}

			// Add pending channel before registering cancellation because the delegate may fire immediately
			// if the cancellation is already cancelled, and that will remove the channelId from the pendingChannels.
			var pendingChannel = new PendingChannel(openMessage, completionSource);
			this.pendingChannels.Add(channelId, pendingChannel);
			pendingChannel.CancellationRegistration = cancellation.Register(() =>
			{
				lock (this.lockObject)
				{
					if (this.pendingChannels.Remove(channelId))
					{
						pendingChannel.CancellationRegistration.Dispose();
						completionSource.TrySetCanceled(cancellation);
					}
				}
			});
		}

		await Session.SendMessageAsync(
			openMessage,
			cancellation).ConfigureAwait(false);

		return channelId;
	}

	internal async Task<SshChannel> AcceptChannelAsync(
		string channelType,
		CancellationToken cancellation)
	{
		var completionSource = new TaskCompletionSource<SshChannel>(
			TaskCreationOptions.RunContinuationsAsynchronously);

		CancellationTokenRegistration cancellationRegistration;
		SshChannel? channel = null;
		lock (this.lockObject)
		{
			if (this.disposed)
			{
				throw this.closedException ??
					new ObjectDisposedException(Session.ToString(), "Session closed.");
			}

			channel = this.nonAcceptedChannels.Values
				.FirstOrDefault(c => c.ChannelType == channelType);
			if (channel != null)
			{
				// Found a channel that was already opened but not accepted.
				this.nonAcceptedChannels.Remove(channel.ChannelId);
				return channel;
			}

			// Set up the completion source to wait for a channel of the requested type.
			if (!this.pendingAcceptChannels.TryGetValue(channelType, out var list))
			{
				list = new List<TaskCompletionSource<SshChannel>>();
				this.pendingAcceptChannels.Add(channelType, list);
			}

			list.Add(completionSource);

			// Register cancellation delegate after adding to pendingAcceptChannels because if it is already cancelled,
			// the delegate will fire immediately and that will remove the completion source from pendingAcceptChannels.
			cancellationRegistration = cancellation.Register(() =>
			{
				lock (this.lockObject)
				{
					if (this.pendingAcceptChannels.TryGetValue(channelType, out var list))
					{
						list.Remove(completionSource);
						if (list.Count == 0)
						{
							this.pendingAcceptChannels.Remove(channelType);
						}
					}
				}

				completionSource.TrySetCanceled(cancellation);
			});
		}

		try
		{
			return await completionSource.Task.ConfigureAwait(false);
		}
		finally
		{
			cancellationRegistration.Dispose();
		}
	}

	internal Task HandleMessageAsync(
		ConnectionMessage message, CancellationToken cancellation)
	{
		if (message == null) throw new ArgumentNullException(nameof(message));

		return message switch
		{
			ChannelDataMessage m => HandleMessageAsync(m, cancellation),
			ChannelWindowAdjustMessage m => HandleMessageAsync(m, cancellation),
			ChannelOpenMessage m => HandleMessageAsync(m, cancellation),
			ChannelOpenConfirmationMessage m => HandleMessageAsync(m, cancellation),
			ChannelOpenFailureMessage m => HandleMessageAsync(m, cancellation),
			ChannelRequestMessage m => HandleMessageAsync(m, cancellation),
			ChannelSuccessMessage m => HandleMessageAsync(m, cancellation),
			ChannelFailureMessage m => HandleMessageAsync(m, cancellation),
			ChannelEofMessage m => HandleMessageAsync(m, cancellation),
			ChannelCloseMessage m => HandleMessageAsync(m, cancellation),
			_ => Task.CompletedTask, // Ignore unrecognized connection messages.
		};
	}

	private async Task HandleMessageAsync(
		ChannelOpenMessage message, CancellationToken cancellation)
	{
		var senderChannel = message.SenderChannel;

		if (!Session.CanAcceptRequests)
		{
			Session.Trace.TraceEvent(
				TraceEventType.Warning,
				SshTraceEventIds.ChannelOpenFailed,
				"Channel open request blocked because the session is not yet authenticated.");
			await Session.SendMessageAsync(
				new ChannelOpenFailureMessage
				{
					RecipientChannel = senderChannel,
					ReasonCode = SshChannelOpenFailureReason.AdministrativelyProhibited,
					Description = "Authenticate before opening channels.",
				},
				cancellation).ConfigureAwait(false);
			return;
		}

		uint channelId = (uint)Interlocked.Increment(ref this.channelCounter);
		if (channelId > int.MaxValue)
		{
			this.channelCounter = int.MaxValue;

			await Session.SendMessageAsync(
				new ChannelOpenFailureMessage
				{
					RecipientChannel = senderChannel,
					ReasonCode = SshChannelOpenFailureReason.ResourceShortage,
					Description = "Maximum number of channels reached.",
				},
				cancellation).ConfigureAwait(false);
			return;
		}

		if (message.ChannelType == null)
		{
			await Session.SendMessageAsync(
				new ChannelOpenFailureMessage
				{
					RecipientChannel = senderChannel,
					ReasonCode = SshChannelOpenFailureReason.UnknownChannelType,
					Description = "Channel type not specified.",
				},
				cancellation).ConfigureAwait(false);
			return;
		}

		// Save a copy of the message because its buffer will be overwritten by the next receive.
		message = message.ConvertTo<ChannelOpenMessage>(copy: true);

		// The confirmation message may be reassigned if the opening task returns a custom message.
		var confirmationMessage = new ChannelOpenConfirmationMessage();

		var channel = new SshChannel(
			this,
			message.ChannelType!,
			channelId: channelId,
			remoteChannelId: senderChannel,
			remoteMaxWindowSize: message.MaxWindowSize,
			remoteMaxPacketSize: message.MaxPacketSize,
			message,
			confirmationMessage);

		ChannelMessage responseMessage;
		var e = new SshChannelOpeningEventArgs(message, channel, isRemoteRequest: true);
		try
		{
			await Session.OnChannelOpeningAsync(e, cancellation).ConfigureAwait(false);
			if (e.OpeningTask != null)
			{
				responseMessage = await e.OpeningTask.ConfigureAwait(false);
			}
			else if (e.FailureReason != SshChannelOpenFailureReason.None)
			{
				responseMessage = new ChannelOpenFailureMessage
				{
					ReasonCode = e.FailureReason,
					Description = e.FailureDescription,
				};
			}
			else
			{
				responseMessage = confirmationMessage;
			}
		}
		catch (ArgumentException aex)
		{
			responseMessage = new ChannelOpenFailureMessage
			{
				ReasonCode = SshChannelOpenFailureReason.ConnectFailed,
				Description = aex.Message,
			};
		}
		catch (Exception)
		{
			channel.Dispose();
			throw;
		}

		if (responseMessage is ChannelOpenFailureMessage)
		{
			responseMessage.RecipientChannel = senderChannel;
			try
			{
				await Session.SendMessageAsync(responseMessage, cancellation).ConfigureAwait(false);
				return;
			}
			finally
			{
				channel.Dispose();
			}
		}

		// Prevent any changes to the channel max window size after sending the value in the
		// open confirmation message.
		channel.IsMaxWindowSizeLocked = true;

		lock (this.lockObject)
		{
			this.channels.Add(channel.ChannelId, channel);
		}

		if (this.disposed)
		{
			// A lot can happen while opening a channel (as when channels route to other services).
			// Ensure the channel gets cleaned up if this object was already disposed.
			channel.Dispose();
			return;
		}

		confirmationMessage = (ChannelOpenConfirmationMessage)responseMessage;
		confirmationMessage.RecipientChannel = channel.RemoteChannelId;
		confirmationMessage.SenderChannel = channel.ChannelId;
		confirmationMessage.MaxWindowSize = channel.MaxWindowSize;
		confirmationMessage.MaxPacketSize = channel.MaxPacketSize;
		confirmationMessage.Rewrite();

		channel.OpenConfirmationMessage = confirmationMessage;
		await Session.SendMessageAsync(confirmationMessage, cancellation).ConfigureAwait(false);

		lock (this.lockObject)
		{
			// Check if there are any accept operations waiting on this channel type.
			bool accepted = false;
			if (this.pendingAcceptChannels.TryGetValue(channel.ChannelType, out var list))
			{
				while (list.Count > 0)
				{
					var acceptCompletionSource = list[0];
					list.RemoveAt(0);

					if (acceptCompletionSource.TrySetResult(channel))
					{
						accepted = true;
						break;
					}
				}

				if (list.Count == 0)
				{
					this.pendingAcceptChannels.Remove(channel.ChannelType);
				}
			}

			if (!accepted)
			{
				this.nonAcceptedChannels.Add(channel.ChannelId, channel);
			}
		}

		OnChannelOpenCompleted(channel.ChannelId, channel);
		channel.EnableSending();
	}

	private async Task HandleMessageAsync(
		ChannelRequestMessage message, CancellationToken cancellation)
	{
		var channel = TryGetChannelForMessage(message);
		if (channel != null)
		{
			await channel.HandleRequestAsync(message, cancellation).ConfigureAwait(false);
		}
	}

	private Task HandleMessageAsync(ChannelSuccessMessage message, CancellationToken cancellation)
	{
		cancellation.ThrowIfCancellationRequested();
		var channel = TryGetChannelForMessage(message);
		channel?.HandleResponse(true);
		return Task.CompletedTask;
	}

	private Task HandleMessageAsync(ChannelFailureMessage message, CancellationToken cancellation)
	{
		cancellation.ThrowIfCancellationRequested();
		var channel = TryGetChannelForMessage(message);
		channel?.HandleResponse(false);
		return Task.CompletedTask;
	}

	private async Task HandleMessageAsync(
		ChannelDataMessage message, CancellationToken cancellation)
	{
		var channel = TryGetChannelForMessage(message);
		if (channel != null)
		{
			await channel.HandleDataReceivedAsync(message.Data, cancellation).ConfigureAwait(false);
		}
	}

	private Task HandleMessageAsync(
		ChannelWindowAdjustMessage message, CancellationToken cancellation)
	{
		cancellation.ThrowIfCancellationRequested();
		var channel = TryGetChannelForMessage(message);
		channel?.AdjustRemoteWindow(message.BytesToAdd);
		return Task.CompletedTask;
	}

	private Task HandleMessageAsync(ChannelEofMessage message, CancellationToken cancellation)
	{
		cancellation.ThrowIfCancellationRequested();
		var channel = TryFindChannelById(message.RecipientChannel);
		channel?.OnEof();
		return Task.CompletedTask;
	}

	private Task HandleMessageAsync(ChannelCloseMessage message, CancellationToken cancellation)
	{
		cancellation.ThrowIfCancellationRequested();
		var channel = TryFindChannelById(message.RecipientChannel);
		channel?.Close();
		return Task.CompletedTask;
	}

	private async Task HandleMessageAsync(
		ChannelOpenConfirmationMessage message, CancellationToken cancellation)
	{
		SshChannel channel;
		cancellation.ThrowIfCancellationRequested();

		TaskCompletionSource<SshChannel>? completionSource = null;
		ChannelOpenMessage openMessage;
		lock (this.lockObject)
		{
			if (this.pendingChannels.TryGetValue(
				message.RecipientChannel, out var pendingChannel))
			{
				openMessage = pendingChannel.OpenMessage;
				completionSource = pendingChannel.CompletionSource;
				pendingChannel.CancellationRegistration.Dispose();
				this.pendingChannels.Remove(message.RecipientChannel);
			}
			else if (this.channels.ContainsKey(message.RecipientChannel))
			{
				throw new SshConnectionException(
					"Duplicate channel confirmation.", SshDisconnectReason.ProtocolError);
			}
			else
			{
				throw new SshConnectionException(
					"Channel confirmation was not requested.", SshDisconnectReason.ProtocolError);
			}

			// Save a copy of the message because its buffer will be overwritten by the next receive.
			message = message.ConvertTo<ChannelOpenConfirmationMessage>(copy: true);

			channel = new SshChannel(
				this,
				openMessage.ChannelType ?? SshChannel.SessionChannelType,
				channelId: message.RecipientChannel,
				remoteChannelId: message.SenderChannel,
				remoteMaxWindowSize: message.MaxWindowSize,
				remoteMaxPacketSize: message.MaxPacketSize,
				openMessage,
				message);

			// Set the channel max window size property to match the value sent in the open message,
			// and lock it to prevent any further changes.
			channel.MaxWindowSize = openMessage.MaxWindowSize;
			channel.IsMaxWindowSizeLocked = true;

			this.channels.Add(channel.ChannelId, channel);
		}

		var args = new SshChannelOpeningEventArgs(openMessage, channel, isRemoteRequest: false);
		await Session.OnChannelOpeningAsync(args, cancellation).ConfigureAwait(false);

		if (completionSource != null)
		{
			if (args.FailureReason == SshChannelOpenFailureReason.None)
			{
				completionSource.TrySetResult(channel);
			}
			else
			{
				completionSource.TrySetException(new SshChannelException(
					args.FailureDescription ?? "Channel open failure: " + args.FailureReason,
					args.FailureReason));
				return;
			}
		}
		else if (args.FailureReason == SshChannelOpenFailureReason.None)
		{
			OnChannelOpenCompleted(channel.ChannelId, channel);
		}

		channel.EnableSending();
	}

	private Task HandleMessageAsync(
		ChannelOpenFailureMessage message, CancellationToken cancellation)
	{
		cancellation.ThrowIfCancellationRequested();

		TaskCompletionSource<SshChannel>? completionSource = null;
		lock (this.lockObject)
		{
			if (this.pendingChannels.TryGetValue(
				message.RecipientChannel, out var pendingChannel))
			{
				completionSource = pendingChannel.CompletionSource;
				pendingChannel.CancellationRegistration.Dispose();
				this.pendingChannels.Remove(message.RecipientChannel);
			}
		}

		if (completionSource != null)
		{
			completionSource.TrySetException(new SshChannelException(
				message.Description ?? "Channel open failure: " + message.ReasonCode,
				message.ReasonCode));
		}
		else
		{
			OnChannelOpenCompleted(message.RecipientChannel, null);
		}

		return Task.CompletedTask;
	}

	private void OnChannelOpenCompleted(uint channelId, SshChannel? channel)
	{
		if (channel != null)
		{
			Session.Trace.TraceEvent(
				TraceEventType.Verbose,
				SshTraceEventIds.ChannelOpened,
				$"{Session} {nameof(OnChannelOpenCompleted)}({channel})");
		}
		else
		{
			Session.Trace.TraceEvent(
				TraceEventType.Verbose,
				SshTraceEventIds.ChannelOpenFailed,
				$"{Session} {nameof(OnChannelOpenCompleted)}({channelId} failed)");
		}
	}

	/// <summary>
	/// Gets the channel object based on the message <see cref="ChannelMessage.RecipientChannel" />
	/// property. Logs a warning if the channel was not found.
	/// </summary>
	private SshChannel? TryGetChannelForMessage(ChannelMessage channelMessage)
	{
		var channel = TryFindChannelById(channelMessage.RecipientChannel);
		if (channel == null)
		{
			string messageString = (channelMessage is ChannelDataMessage) ?
				nameof(ChannelDataMessage) : channelMessage.ToString();
			Trace.TraceEvent(
				TraceEventType.Warning,
				SshTraceEventIds.ChannelRequestFailed,
				$"{Session}: Invalid channel ID {channelMessage.RecipientChannel} in {messageString}");
		}

		return channel;
	}

	private SshChannel? TryFindChannelById(uint id)
	{
		lock (this.lockObject)
		{
			if (!this.channels.TryGetValue(id, out SshChannel? channel))
			{
				channel = null;
			}

			return channel;
		}
	}

	internal void RemoveChannel(SshChannel channel)
	{
		lock (this.lockObject)
		{
			this.channels.Remove(channel.ChannelId);
			this.nonAcceptedChannels.Remove(channel.ChannelId);
		}
	}
}

File: src/ts/ssh/services/connectionService.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { SshService } from './sshService';
import {
	ConnectionMessage,
	ChannelOpenMessage,
	ChannelCloseMessage,
	ChannelOpenConfirmationMessage,
	ChannelOpenFailureMessage,
	ChannelWindowAdjustMessage,
	ChannelDataMessage,
	ChannelRequestMessage,
	ChannelSuccessMessage,
	ChannelFailureMessage,
	ChannelEofMessage,
	SshChannelOpenFailureReason,
	ChannelMessage,
	ChannelExtendedDataMessage,
} from '../messages/connectionMessages';
import { SshSession } from '../sshSession';
import { CancellationToken, Disposable } from 'vscode-jsonrpc';
import { PromiseCompletionSource } from '../util/promiseCompletionSource';
import { SshChannel } from '../sshChannel';
import { CancellationError } from '../util/cancellation';
import { ObjectDisposedError, SshChannelError } from '../errors';
import { SshChannelOpeningEventArgs } from '../events/sshChannelOpeningEventArgs';
import { serviceActivation } from './serviceActivation';
import { TraceLevel, SshTraceEventIds } from '../trace';
import { SshExtendedDataEventArgs } from '../events/sshExtendedDataEventArgs';

interface PendingChannel {
	openMessage: ChannelOpenMessage;
	completionSource: PromiseCompletionSource<SshChannel>;
	cancellationRegistration: Disposable | null;
}

@serviceActivation({ serviceRequest: ConnectionService.serviceName })
export class ConnectionService extends SshService {
	public static readonly serviceName = 'ssh-connection';

	private channelCounter: number = 0;

	private readonly channelMap = new Map<number, SshChannel>();
	private readonly nonAcceptedChannels = new Map<number, SshChannel>();
	private readonly pendingChannels = new Map<number, PendingChannel>();
	private readonly pendingAcceptChannels = new Map<
		string,
		PromiseCompletionSource<SshChannel>[]
	>();

	public constructor(session: SshSession) {
		super(session);
	}

	public get channels(): SshChannel[] {
		return Array.from(this.channelMap.values());
	}

	public close(e: Error): void {
		let channelCompletions = [...this.pendingChannels.values()].map((pc) => pc.completionSource);
		if (this.pendingAcceptChannels.size > 0) {
			channelCompletions = channelCompletions.concat(
				[...this.pendingAcceptChannels.values()].reduce((a, b) => a.concat(b)),
			);
		}

		for (const channel of this.channelMap.values()) {
			channel.close(e);
		}

		for (const channelCompletion of channelCompletions) {
			channelCompletion.reject(e);
		}
	}

	public dispose(): void {
		const channels = [...this.channelMap.values()];

		let channelCompletions = [...this.pendingChannels.values()].map((pc) => pc.completionSource);
		if (this.pendingAcceptChannels.size > 0) {
			channelCompletions = channelCompletions.concat(
				[...this.pendingAcceptChannels.values()].reduce((a, b) => a.concat(b)),
			);
		}

		for (const channel of channels) {
			channel.dispose();
		}

		for (const channelCompletion of channelCompletions) {
			channelCompletion.reject(new ObjectDisposedError('Session closed.'));
		}

		super.dispose();
	}

	public async acceptChannel(
		channelType: string,
		cancellation?: CancellationToken,
	): Promise<SshChannel> {
		const completionSource = new PromiseCompletionSource<SshChannel>();

		let cancellationRegistration: Disposable | undefined;
		if (cancellation) {
			if (cancellation.isCancellationRequested) throw new CancellationError();
			cancellationRegistration = cancellation.onCancellationRequested(() => {
				const list = this.pendingAcceptChannels.get(channelType);
				if (list) {
					const index = list.findIndex((item) => Object.is(item, completionSource));
					if (index >= 0) {
						list.splice(index, 1);
					}
				}

				completionSource.reject(new CancellationError());
			});
		}

		let channel: SshChannel | null = null;

		channel =
			Array.from(this.nonAcceptedChannels.values()).find((c) => c.channelType === channelType) ||
			null;
		if (channel) {
			// Found a channel that was already opened but not accepted.
			this.nonAcceptedChannels.delete(channel.channelId);
		} else {
			// Set up the completion source to wait for a channel of the requested type.
			let list = this.pendingAcceptChannels.get(channelType);
			if (!list) {
				list = [];
				this.pendingAcceptChannels.set(channelType, list);
			}

			list.push(completionSource);
		}

		try {
			return channel || (await completionSource.promise);
		} finally {
			if (cancellationRegistration) cancellationRegistration.dispose();
		}
	}

	public async openChannel(
		openMessage: ChannelOpenMessage,
		completionSource: PromiseCompletionSource<SshChannel>,
		cancellation?: CancellationToken,
	): Promise<number> {
		const channelId = ++this.channelCounter;
		openMessage.senderChannel = channelId;

		let cancellationRegistration: Disposable | null = null;
		if (cancellation) {
			if (cancellation.isCancellationRequested) throw new CancellationError();
			cancellationRegistration = cancellation.onCancellationRequested(() => {
				if (this.pendingChannels.delete(channelId)) {
					completionSource.reject(new CancellationError());
				}
			});
		}

		this.pendingChannels.set(channelId, {
			openMessage: openMessage,
			completionSource: completionSource,
			cancellationRegistration: cancellationRegistration,
		});

		await this.session.sendMessage(openMessage);

		return channelId;
	}

	public handleMessage(
		message: ConnectionMessage,
		cancellation?: CancellationToken,
	): void | Promise<void> {
		if (message instanceof ChannelDataMessage) {
			return this.handleDataMessage(message);
		} else if (message instanceof ChannelExtendedDataMessage) {
			return this.handleExtendedDataMessage(message);
		} else if (message instanceof ChannelWindowAdjustMessage) {
			return this.handleAdjustWindowMessage(message);
		} else if (message instanceof ChannelEofMessage) {
			return this.handleEofMessage(message);
		} else if (message instanceof ChannelOpenMessage) {
			return this.handleOpenMessage(message, cancellation);
		} else if (message instanceof ChannelCloseMessage) {
			return this.handleCloseMessage(message);
		} else if (message instanceof ChannelOpenConfirmationMessage) {
			return this.handleOpenConfirmationMessage(message, cancellation);
		} else if (message instanceof ChannelOpenFailureMessage) {
			return this.handleOpenFailureMessage(message);
		} else if (message instanceof ChannelRequestMessage) {
			return this.handleRequestMessage(message, cancellation);
		} else if (message instanceof ChannelSuccessMessage) {
			return this.handleSuccessMessage(message);
		} else if (message instanceof ChannelFailureMessage) {
			return this.handleFailureMessage(message);
		} else {
			throw new Error(`Message not implemented: ${message}`);
		}
	}

	private async handleOpenMessage(
		message: ChannelOpenMessage,
		cancellation?: CancellationToken,
	): Promise<void> {
		const senderChannel = message.senderChannel;
		if (!this.session.canAcceptRequests) {
			this.trace(
				TraceLevel.Warning,
				SshTraceEventIds.channelOpenFailed,
				'Channel open request blocked because the session is not yet authenticated.',
			);
			const openFailureMessage = new ChannelOpenFailureMessage();
			openFailureMessage.recipientChannel = senderChannel;
			openFailureMessage.reasonCode = SshChannelOpenFailureReason.administrativelyProhibited;
			openFailureMessage.description = 'Authenticate before opening channels.';
			await this.session.sendMessage(openFailureMessage, cancellation);
			return;
		} else if (!message.channelType) {
			const openFailureMessage = new ChannelOpenFailureMessage();
			openFailureMessage.recipientChannel = senderChannel;
			openFailureMessage.reasonCode = SshChannelOpenFailureReason.unknownChannelType;
			openFailureMessage.description = 'Channel type not specified.';
			await this.session.sendMessage(openFailureMessage, cancellation);
			return;
		}

		// Save a copy of the message because its buffer will be overwitten by the next receive.
		message = message.convertTo(new ChannelOpenMessage(), true);

		// The confirmation message may be reassigned if the opening task returns a custom message.
		let confirmationMessage = new ChannelOpenConfirmationMessage();

		const channelId = ++this.channelCounter;
		const channel = new SshChannel(
			this,
			message.channelType!,
			channelId,
			senderChannel!,
			message.maxWindowSize!,
			message.maxPacketSize!,
			message,
			confirmationMessage,
		);

		let responseMessage: ChannelMessage;
		const args = new SshChannelOpeningEventArgs(message, channel, true);
		try {
			await this.session.handleChannelOpening(args, cancellation);
			if (args.openingPromise) {
				responseMessage = await args.openingPromise;
			} else if (args.failureReason !== SshChannelOpenFailureReason.none) {
				const failureMessage = new ChannelOpenFailureMessage();
				failureMessage.reasonCode = args.failureReason;
				failureMessage.description = args.failureDescription ?? undefined;
				responseMessage = failureMessage;
			} else {
				responseMessage = confirmationMessage;
			}
		} catch (e) {
			channel.dispose();
			throw e;
		}

		if (responseMessage instanceof ChannelOpenFailureMessage) {
			responseMessage.recipientChannel = senderChannel;
			try {
				await this.session.sendMessage(responseMessage, cancellation);
			} finally {
				channel.dispose();
			}
			return;
		}

		// The session might have been closed while opening the channel.
		if (this.session.isClosed) {
			channel.dispose();
			return;
		}

		// Prevent any changes to the channel max window size after sending the value in the
		// open confirmation message.
		channel.isMaxWindowSizeLocked = true;

		this.channelMap.set(channel.channelId, channel);

		confirmationMessage = <ChannelOpenConfirmationMessage>responseMessage;
		confirmationMessage.recipientChannel = channel.remoteChannelId;
		confirmationMessage.senderChannel = channel.channelId;
		confirmationMessage.maxWindowSize = channel.maxWindowSize;
		confirmationMessage.maxPacketSize = channel.maxPacketSize;
		confirmationMessage.rewrite();

		channel.openConfirmationMessage = confirmationMessage;
		await this.session.sendMessage(confirmationMessage, cancellation);

		// Check if there are any accept operations waiting on this channel type.
		let accepted = false;
		const list = this.pendingAcceptChannels.get(channel.channelType);
		while (list && list.length > 0) {
			const acceptCompletionSource = list.shift();
			acceptCompletionSource!.resolve(channel);
			accepted = true;
			break;
		}

		if (!accepted) {
			this.nonAcceptedChannels.set(channel.channelId, channel);
		}

		this.onChannelOpenCompleted(channel.channelId, channel);
		channel.enableSending();
	}

	private handleCloseMessage(message: ChannelCloseMessage): void {
		const channel = this.findChannelById(message.recipientChannel!);
		if (channel) {
			channel.handleClose();
		}
	}

	private async handleOpenConfirmationMessage(
		message: ChannelOpenConfirmationMessage,
		cancellation?: CancellationToken,
	): Promise<void> {
		let completionSource: PromiseCompletionSource<SshChannel> | null = null;

		let openMessage: ChannelOpenMessage;
		const pendingChannel = this.pendingChannels.get(message.recipientChannel!);
		if (pendingChannel) {
			openMessage = pendingChannel.openMessage;
			completionSource = pendingChannel.completionSource;
			if (pendingChannel.cancellationRegistration) {
				pendingChannel.cancellationRegistration.dispose();
			}
			this.pendingChannels.delete(message.recipientChannel!);
		} else if (this.channelMap.has(message.recipientChannel!)) {
			throw new Error('Duplicate channel ID.');
		} else {
			throw new Error('Channel confirmation was not requested.');
		}

		// Save a copy of the message because its buffer will be overwitten by the next receive.
		message = message.convertTo(new ChannelOpenConfirmationMessage(), true);

		const channel = new SshChannel(
			this,
			openMessage.channelType || SshChannel.sessionChannelType,
			message.recipientChannel!,
			message.senderChannel!,
			message.maxWindowSize!,
			message.maxPacketSize!,
			openMessage,
			message,
		);

		// Set the channel max window size property to match the value sent in the open message,
		// (if specified) and lock it to prevent any further changes.
		if (typeof openMessage.maxWindowSize === 'number') {
			channel.maxWindowSize = openMessage.maxWindowSize;
		}
		channel.isMaxWindowSizeLocked = true;

		this.channelMap.set(channel.channelId, channel);

		const args = new SshChannelOpeningEventArgs(openMessage, channel, false);
		await this.session.handleChannelOpening(args, cancellation);

		if (completionSource) {
			if (args.failureReason === SshChannelOpenFailureReason.none) {
				completionSource.resolve(channel);
			} else {
				completionSource.reject(
					new SshChannelError(
						args.failureDescription ?? 'Channel open failure.',
						args.failureReason,
					),
				);
				return;
			}
		} else {
			this.onChannelOpenCompleted(channel.channelId, channel);
		}

		channel.enableSending();
	}

	private handleOpenFailureMessage(message: ChannelOpenFailureMessage): void {
		let completionSource: PromiseCompletionSource<SshChannel> | null = null;

		const pendingChannel = this.pendingChannels.get(message.recipientChannel!);
		if (pendingChannel) {
			completionSource = pendingChannel.completionSource;
			if (pendingChannel.cancellationRegistration) {
				pendingChannel.cancellationRegistration.dispose();
			}
			this.pendingChannels.delete(message.recipientChannel!);
		}

		if (completionSource != null) {
			completionSource.reject(
				new SshChannelError(
					message.description || 'Channel open rejected.',
					message.reasonCode,
				),
			);
		} else {
			this.onChannelOpenCompleted(message.recipientChannel!, null);
		}
	}

	private async handleRequestMessage(
		message: ChannelRequestMessage,
		cancellation?: CancellationToken,
	): Promise<void> {
		const channel = this.tryGetChannelForMessage(message);
		if (!channel) return;

		await channel.handleRequest(message, cancellation);
	}

	private handleSuccessMessage(message: ChannelSuccessMessage) {
		const channel = this.tryGetChannelForMessage(message);
		channel?.handleResponse(true);
	}

	private handleFailureMessage(message: ChannelFailureMessage) {
		const channel = this.tryGetChannelForMessage(message);
		channel?.handleResponse(false);
	}

	private handleDataMessage(message: ChannelDataMessage): void {
		const channel = this.tryGetChannelForMessage(message);
		channel?.handleDataReceived(message.data!);
	}

	private handleExtendedDataMessage(message: ChannelExtendedDataMessage): void {
		const channel = this.tryGetChannelForMessage(message);
		channel?.handleExtendedDataReceived(new SshExtendedDataEventArgs(message.dataTypeCode!, message.data!));
	}

	private handleAdjustWindowMessage(message: ChannelWindowAdjustMessage): void {
		const channel = this.tryGetChannelForMessage(message);
		channel?.adjustRemoteWindow(message.bytesToAdd!);
	}

	private handleEofMessage(message: ChannelEofMessage): void {
		const channel = this.findChannelById(message.recipientChannel!);
		channel?.handleEof();
	}

	private onChannelOpenCompleted(channelId: number, channel: SshChannel | null) {
		if (channel) {
			this.trace(
				TraceLevel.Info,
				SshTraceEventIds.channelOpened,
				`${this.session} ChannelOpenCompleted(${channel})`,
			);
		} else {
			this.trace(
				TraceLevel.Warning,
				SshTraceEventIds.channelOpenFailed,
				`${this.session} ChannelOpenCompleted(${channelId} failed)`,
			);
		}
	}

	/**
	 * Gets the channel object based on the message `recipientChannel` property.
	 * Logs a warning if the channel was not found.
	 */
	private tryGetChannelForMessage(channelMessage: ChannelMessage): SshChannel | null {
		const channel = this.findChannelById(channelMessage.recipientChannel!);
		if (!channel) {
			const messageString =
				channelMessage instanceof ChannelDataMessage
					? 'channel data message'
					: channelMessage.toString();
			this.trace(
				TraceLevel.Warning,
				SshTraceEventIds.channelRequestFailed,
				`Invalid channel ID ${channelMessage.recipientChannel} in ${messageString}.`,
			);
		}
		return channel;
	}

	private findChannelById(id: number): SshChannel | null {
		const channel = this.channelMap.get(id) ?? null;
		return channel;
	}

	/* @internal */
	public removeChannel(channel: SshChannel) {
		this.channelMap.delete(channel.channelId);
		this.pendingChannels.delete(channel.channelId);
	}
}
